Feature/structured logging

dpetran opened the pull request.

dpetran committed:
add INFO structured logs

The INFO log level records user actions and contains no internal datastructures.

dpetran committed:
add DEBUG structured logs

The DEBUG log level records application logic and contains internal datastructures where
relevant.

dpetran committed:
add TRACE structured logs

Right now I'm just logging cache misses.

dpetran committed:
fixup debug

dpetran committed:
add WARN structured logs

WARN is for alerting the administrator of the database of possibly degraded
conditions.

dpetran committed:
add structured ERROR logs

Every existing error log has been converted over.

dpetran committed:
add opentelemetry configuration

dpetran committed:
add memory signal handler

Also convert logging functions to macros to retain callsite info.

dpetran committed:
fix formatting

dpetran committed:
add tracing to info logs

This allows us to correlate logs to a single span. Exceptions thrown within a trace a
wrapped with additional context, so tests expecting exceptions needed some unwrapping.

dpetran commented:
This is just a work-in-progress, the otel stuff is just for improved iteration while we're figuring out how to plumb everything together. Likewise, the separate logging api (with ! suffixes) are to make iterating easier, the plan is to integrate it properly once the basic tracing functionality is working.

dpetran committed:
add tracing to debug logs

dpetran committed:
configure tools.logging to use telemere as its backend

tools.logging calls will now create telemere signals

bplatz commented:
One thing to consider in making it work correctly is making it work like an actual app would - so I think the goal would be to configure the telemere backend there, since we cannot ship telemere as a backend with fluree/db. That way you can make sure that the fluree/db loging flows through to the telemere backend (e.g. in fluree/server) as you expect... and I think the key here is MDC handling which is the tricky part, especially across threads (including core.async which was our issue before). If we want to adopt telemere as the backend logger in fluree/server that is fine, but that is where the backend logger code needs to live. I presume we'd remain with SLF4J as the facade here, but maybe you had a different idea in mind.

dpetran committed:
use trove as structured logging facade

Telemere is a logging implementation, trove is a tiny facade for telemere that supports
any logging backend.

dpetran committed:
use clj-otel for tracing

dpetran committed:
move tracing to separate api

dpetran committed:
add tracing to transaction flows

dpetran committed:
remove experimental logging api

dpetran committed:
add noop cljs impl for tracing
